### PR TITLE
[PLA-1943] Fix relationship total count

### PR DESCRIPTION
**Pull request by enjinabner**
### **PR Type**
Bug fix, Tests

___

### **Description**
- Fixed the calculation of total counts for tokens and accounts in `GetCollectionsQuery` by adding logic to fetch and append these counts to collection items.
- Introduced new database queries to accurately compute the total number of tokens and accounts associated with each collection.
- Added a new test case in `GetCollectionsTest` to ensure the total counts for tokens and accounts are calculated and returned correctly.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>GetCollectionsQuery.php</strong><dd><code>Fix and enhance total count calculation for collections</code>&nbsp; &nbsp; </dd></summary>
<hr>

src/GraphQL/Schemas/Primary/Substrate/Queries/GetCollectionsQuery.php

<li>Added logic to calculate and append <code>tokens_count</code> and <code>accounts_count</code> to <br>collection items.<br> <li> Utilized <code>Arr</code> helper for array operations.<br> <li> Introduced additional database queries to fetch total counts for <br>tokens and accounts.<br>

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-core/pull/228/files#diff-eb7ee036c98547bf0961b1d3ce78fdd4f31518ccc8c90bdb699b8b68c5484e97">+29/-1</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>GetCollectionsTest.php</strong><dd><code>Add test for verifying total count calculation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/Feature/GraphQL/Queries/GetCollectionsTest.php

<li>Added a new test to verify correct total count calculation for tokens <br>and accounts.<br> <li> Utilized <code>Arr</code> helper for assertions in the test.<br> <li> Ensured database is reset before running the new test.<br>

</details>

  </td>
  <td><a href="https://github.com/enjin/platform-core/pull/228/files#diff-3e7f3dd5c92b8e73201c2afeafb7d1d500a88e914560c29b3ac436cc974b08af">+24/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: src/GraphQL/Schemas/Primary/Substrate/Queries/GetCollectionsQuery.php [hunk 1]
```diff
@@ -68,10 +71,35 @@ public function resolve(
...
+                ->select('collection_id', DB::raw('count(*) as total'))
+                ->groupBy('collection_id')
+                ->pluck('total', 'collection_id');
+
+            foreach ($items as $item) {
+                $item->accounts_count = $totalCounts->get($item->id, 0);
+            }
+        }
+
+        return $result;
```
Comment (v16Studios) [inline src/GraphQL/Schemas/Primary/Substrate/Queries/GetCollectionsQuery.php:102]:
```suggestion
        }
        
        $result['items']->setCollection(collect($items));

        return $result;
```
Should the `$items` be assigned back to the `$result` before returning it?  Or did I miss something 😅

File: tests/Feature/GraphQL/Queries/GetCollectionsTest.php [hunk 2]
```diff
@@ -45,6 +46,29 @@ public function test_it_can_fetch_with_no_args(): void
...
+    {
+        $this->artisan('migrate:fresh');
+
+        $collections = Collection::factory(2)->create();
+        $collection1 = $collections->shift();
+        Token::factory(10)->create(['collection_id' => $collection1->id]);
+        CollectionAccount::factory(10)->create(['collection_id' => $collection1->id]);
+        $collection2 = $collections->shift();
+        Token::factory(10)->create(['collection_id' => $collection2->id]);
+        CollectionAccount::factory(10)->create(['collection_id' => $collection2->id]);
```
Comment (v16Studios) [inline tests/Feature/GraphQL/Queries/GetCollectionsTest.php:59]: Should we use bigger numbers in the tests?  The UI was showing no more than 15 even if there were many more items, so I'm guessing that any number up to 15 would likely pass the test regardless.

**Review by v16Studios — CHANGES_REQUESTED** [review]

File: src/GraphQL/Schemas/Primary/Substrate/Queries/GetCollectionsQuery.php (hunk 1 above)
Comment (enjinabner) [reply src/GraphQL/Schemas/Primary/Substrate/Queries/GetCollectionsQuery.php:102]: There's no need, object is passed by reference so altering the values will immediately reflect.

File: tests/Feature/GraphQL/Queries/GetCollectionsTest.php (hunk 2 above)
Comment (enjinabner) [reply tests/Feature/GraphQL/Queries/GetCollectionsTest.php:59]: Yes I can increase, i made it smaller purposely so it doesn't have big impact on the unit tests performance.

File: tests/Feature/GraphQL/Queries/GetCollectionsTest.php [hunk 3]
```diff
@@ -45,6 +47,32 @@ public function test_it_can_fetch_with_no_args(): void
...
+        CollectionAccount::truncate();
+        Schema::enableForeignKeyConstraints();
+
+        $collections = Collection::factory(2)->create();
+        $collection1 = $collections->shift();
+        Token::factory($tokenCount1 = random_int(1, 100))->create(['collection_id' => $collection1->id]);
+        CollectionAccount::factory($accountCount1 = random_int(1, 100))->create(['collection_id' => $collection1->id]);
+        $collection2 = $collections->shift();
+        Token::factory($tokenCount2 = random_int(1, 100))->create(['collection_id' => $collection2->id]);
+        CollectionAccount::factory($accountCount2 = random_int(1, 100))->create(['collection_id' => $collection2->id]);
```
Comment (v16Studios) [inline tests/Feature/GraphQL/Queries/GetCollectionsTest.php:64]:
```suggestion
        Token::factory($tokenCount1 = random_int(16, 100))->create(['collection_id' => $collection1->id]);
        CollectionAccount::factory($accountCount1 = random_int(16, 100))->create(['collection_id' => $collection1->id]);
        $collection2 = $collections->shift();
        Token::factory($tokenCount2 = random_int(16, 100))->create(['collection_id' => $collection2->id]);
        CollectionAccount::factory($accountCount2 = random_int(16, 100))->create(['collection_id' => $collection2->id]);
```
Let's use 16 as the min value seeing as the UI was happy to display up to 15 without issue, so this makes sure we don't get false positives for 1-15 which looks like would always work.

**Review by v16Studios — DISMISSED** [review]

**Review by leonardocustodio — APPROVED** [review]

**Review by leonardocustodio — DISMISSED** [review]

**Comment by leonardocustodio** [comment]
Can you check why the tests fail? I guess it could be because I removed flush from setup of all tests. Can you see if you have to forget some cache on those tests?

**Pull request merged**